src/main/java/org/jenkinsci/plugins/scriptsecurity/sandbox/whitelists/EnumeratingWhitelist.java

```diff
@@ -66,21 +66,24 @@ private void cacheSignatureList(List<Signature> ...sigs) {
         }
     }
 
-    /** Prepopulates the "permitted" cache. */
-    public void precache() {
+    /** Prepopulates the "permitted" cache, resetting if populated already.  Should be called when method signatures change or after initialization. */
+    final void precache() {
+        if (!permittedCache.isEmpty()) {
+            this.permittedCache.clear();  // No sense calling clearCache
+        }
         cacheSignatureList((List)methodSignatures(), (List)(newSignatures()),
                            (List)(staticMethodSignatures()), (List)(fieldSignatures()),
                            (List)(staticFieldSignatures()));
     }
 
-    /** Frees up memory used for the cache. */
-    void clearCache() {
+    /** Frees up nearly all memory used for the cache.  MUST BE CALLED if you change the result of the xxSignatures() methods. */
```

Comment:
But of course no known implementations do so, and as this is package-private no outside implementations could start doing so.

Again I think it would be wise to update the class Javadoc to emphasize that you are not allowed to dynamically change the signatures. If you need that for some weird reason, you would need to extend `Whitelist` directly, or (more likely) extend `ProxyWhitelist` and change the delegates.